airbyte-workers/src/main/java/io/airbyte/workers/DefaultReplicationWorker.java

```diff
@@ -269,23 +279,39 @@ private static Runnable getReplicationRunnable(final AirbyteSource source,
       var recordsRead = 0;
       try {
         while (!cancelled.get() && !source.isFinished()) {
-          final Optional<AirbyteMessage> messageOptional = source.attemptRead();
+          final Optional<AirbyteMessage> messageOptional;
+          try {
+            messageOptional = source.attemptRead();
+          } catch (final Exception e) {
+            throw new SourceException("Source process read attempt failed", e);
+          }
           if (messageOptional.isPresent()) {
             final AirbyteMessage message = mapper.mapMessage(messageOptional.get());
 
             messageTracker.acceptFromSource(message);
-            destination.accept(message);
+            try {
+              destination.accept(message);
+            } catch (final Exception e) {
+              throw new DestinationException("Destination process message delivery failed", e);
+            }
             recordsRead += 1;
 
             if (recordsRead % 1000 == 0) {
               LOGGER.info("Records read: {}", recordsRead);
             }
           }
         }
-        destination.notifyEndOfStream();
+        try {
+          destination.notifyEndOfStream();
+        } catch (final Exception e) {
+          throw new DestinationException("Destination process end of stream notification failed", e);
+        }
         if (!cancelled.get() && source.getExitValue() != 0) {
-          throw new RuntimeException("Source process exited with non-zero exit code " + source.getExitValue());
+          throw new SourceException("Source process exited with non-zero exit code " + source.getExitValue());
         }
+      } catch (final SourceException | DestinationException e) {
+        // Surface source and destination exceptions to caller so they can be classified properly
+        throw e;
```

Comment:
Should these be thrown inside of the `if (!cancelled.get())` statement in the below `catch`? What are the consequences of throwing these errors here if we are in the cancellation case?